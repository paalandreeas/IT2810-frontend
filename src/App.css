:root {
  --footer-height: 100px;
}

.App {
  min-height: calc(100vh - var(--footer-height));
}

#logo {
  width: 70px;
  cursor: pointer;
}

.grow {
  flex-grow: 1;
}

#footer {
  height: var(--footer-height);
  margin: 0;
}

#footerContent {
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 100%;
  padding: 0 20px;
  text-align: center;
}
