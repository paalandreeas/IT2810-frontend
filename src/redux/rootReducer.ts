import { combineReducers } from "redux";
import searchReducer from "./search/searchReducer";
import filterReducer from "./filter/filterReducer";
import sortReducer from "./sort/sortReducer";
import movieReducer from "./movie/movieReducer";
import userReducer from "./user/userReducer";
import reviewReducer from "./review/reviewReducer";
import alertReducer from "./alert/alertReducer";

// Creates a root reducer from all the reducers.
const rootReducer = combineReducers({
  search: searchReducer,
  filter: filterReducer,
  sort: sortReducer,
  movieInfo: movieReducer,
  userInfo: userReducer,
  reviewInfo: reviewReducer,
  alertInfo: alertReducer,
});

export default rootReducer;
