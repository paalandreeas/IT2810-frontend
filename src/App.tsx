import React from "react";
import {
  AppBar,
  Box,
  createMuiTheme,
  IconButton,
  makeStyles,
  MuiThemeProvider,
  Toolbar,
} from "@material-ui/core";
import "./App.css";
import { Brightness5, Brightness7 } from "@material-ui/icons";
import AccountCircleIcon from "@material-ui/icons/AccountCircle";
import { useHistory } from "react-router-dom";
import SwitchContainer from "./components/SwitchContainer";
import { RootState } from "./interfaces/RootState";
import { connect, ConnectedProps } from "react-redux";
import CustomSnackbar from "./components/CustomSnackbar";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { resetSearch } from "./redux/search/searchActions";
import { resetFilters } from "./redux/filter/filterActions";

/* 

This first part of the code is all regarding to redux. This is because this component needs to access the redux state.
The following code follows the conventions that redux suggests. You can read more about it here: https://redux.js.org/recipes/usage-with-typescript

The first thing to do is declare an interface (RootState) that fits the redux store. It is only necessary to declare the parts of the store that the component needs.
The function mapStateToProps is responsible for mapping the redux state to the components props. This uses the RootState interface to get the parts that we need from the state.

The function mapDispatchToProps is responsible for mapping the redux actions to the components props. We only map the actions that this component needs and uses.

Then we declare the connector with the connect function with the mapStateToProps and mapDispatchToProps as input.
This is to easily Use the ConnectedProps<T> to infer the types of the props from connect automatically.

Then we declare the type of the final Props that the component will use. We write it like this so that it is easy to add props if the component needs it.

Finally the component takes in props of the type Props.

When the component is exported (at the bottom), the component gets connected to the redux store with the connector we declared.

*/

const mapStateToProps = (state: RootState) => {
  return {
    userInfo: state.userInfo,
  };
};

const mapDispatchToProps = (dispatch: ThunkDispatch<any, any, AnyAction>) => {
  return {
    resetSearch: () => dispatch(resetSearch()),
    resetFilters: () => dispatch(resetFilters()),
  };
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux;

const App: React.FunctionComponent<Props> = (props) => {
  // State to hold the value for darkmade. Set to true by default.
  const [darkMode, setDarkMode] = React.useState(true);

  const primaryColors = {
    main: "#7cec9f",
    light: "#b0ffd0",
    dark: "#47b970",
  };

  // Creates the lightTheme for our website with primary and secondary colors. Part of Material UI.
  const lightTheme = createMuiTheme({
    palette: {
      primary: primaryColors,
      secondary: {
        main: "#e8e8e8",
        light: "#ffffff",
        dark: "#b6b6b6",
      },
      success: primaryColors,
      type: "light",
    },
  });

  // Creates the lightTheme for our website with primary and secondary colors. Part of Material UI.
  const darkTheme = createMuiTheme({
    palette: {
      primary: primaryColors,
      secondary: {
        main: "#212121",
        light: "#484848",
        dark: "#000000",
      },
      success: primaryColors,
      type: "dark",
    },
  });
  // Primary colors are the same in both darkTheme and lightTheme.

  // useMemo recomputes the current theme when one of the dependencies changes.
  // lightTheme and darkTheme never changes.
  // Only when the state darkMode changes, the theme will be recomputed based on that value.
  // If true, use darkmode. Else use lightMode
  const theme = React.useMemo(() => (darkMode ? darkTheme : lightTheme), [
    darkMode,
    darkTheme,
    lightTheme,
  ]);

  // Use makeStyles from MUI to overwrite the styling of MUI components
  // Will only be used on the appBar component, but written for possibility for expansion.
  const useStyles = makeStyles(() => ({
    // Want the appBars bgColor to be the light version of the current theme.
    appBar: { backgroundColor: theme.palette.secondary.light },
  }));

  // Make the classes from useStyles.
  const classes = useStyles();

  // The useHistory hook gives access to the history instance used for navigation.
  const history = useHistory();

  return (
    //  Wraps the entire application in the MUI theme provider so that every component can access the current theme.
    <MuiThemeProvider theme={theme}>
      {/* Box works as a div, but has the possibility to dynamically change bg and text color based on the current theme. Imported from MUI */}
      <Box className="App" bgcolor="secondary.main" color="text.secondary">
        {/* AppBar is imported from MUI. Fixed to the top of the screen. 
                    Makes use of the class "appBar" created with makeStyles. This overwrites the standard bgColor.*/}
        <AppBar position="fixed" className={classes.appBar}>
          {/* AppBar should always be used with the Toolbar component from MUI */}
          <Toolbar>
            {/* the logo of the website */}
            <img
              id="logo"
              src="./resources/images/logo.png"
              alt="logo"
              onClick={() => {
                history.replace("/");
                props.resetSearch();
                props.resetFilters();
              }}
            />
            {/* div that occupies remaining space of the Toolbar
                            This is to space the IconButton to the right side of the toolbar */}
            <div className={"grow"} />
            {/* IconButton from MUI. When clicked, flips the darkMode state.
                            Current icon is based on the darkMode state */}
            <IconButton onClick={() => setDarkMode(!darkMode)}>
              {darkMode ? <Brightness5 /> : <Brightness7 />}
            </IconButton>
            <IconButton
              onClick={() => {
                props.userInfo.loggedIn
                  ? history.push("/user/" + props.userInfo.user.userID)
                  : history.push("/login");
              }}
            >
              <AccountCircleIcon />
            </IconButton>
          </Toolbar>
        </AppBar>
        <SwitchContainer />
      </Box>
      <CustomSnackbar />
      <Box
        id="footer"
        bgcolor="secondary.light"
        color="text.secondary"
        boxShadow={3}
      >
        <div id="footerContent">
          <span>Made with love {"<3"}</span>
        </div>
      </Box>
    </MuiThemeProvider>
  );
};

export default connector(App);
