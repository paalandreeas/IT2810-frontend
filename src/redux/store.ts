import { applyMiddleware, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from "redux-thunk";
import rootReducer from "./rootReducer";

// Creates a redux store from the rootReducer and applies devTools and thunk middleware
const store = createStore(
    rootReducer,
    composeWithDevTools(applyMiddleware(thunk))
);

export default store;
