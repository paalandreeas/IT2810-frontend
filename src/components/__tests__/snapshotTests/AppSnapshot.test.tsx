import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import App from "../../../App";
import { applyMiddleware, createStore } from "redux";
import rootReducer from "../../../redux/rootReducer";
import thunk from "redux-thunk";
import { BrowserRouter as Router } from "react-router-dom";

//creates a store for the provider
const store = createStore(rootReducer, applyMiddleware(thunk));

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
