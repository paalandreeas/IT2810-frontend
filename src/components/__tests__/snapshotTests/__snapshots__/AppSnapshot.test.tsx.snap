// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="MuiBox-root MuiBox-root-2 App"
  >
    <header
      className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-1 mui-fixed MuiPaper-elevation4"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <img
          alt="logo"
          id="logo"
          onClick={[Function]}
          src="./resources/images/logo.png"
        />
        <div
          className="grow"
        />
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 15.31L23.31 12 20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </span>
        </button>
      </div>
    </header>
    <div
      id="switchContainer"
    >
      <div
        id="searchContainer"
      >
        <div
          id="searchBar"
        >
          <div
            id="sortSelect"
          >
            <div
              className="MuiFormControl-root MuiFormControl-fullWidth"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                data-shrink={true}
              >
                Sort by:
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root select MuiInputBase-formControl"
                onClick={[Function]}
              >
                <div
                  aria-haspopup="listbox"
                  className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  Title
                </div>
                <input
                  aria-hidden={true}
                  className="MuiSelect-nativeInput"
                  onAnimationStart={[Function]}
                  onChange={[Function]}
                  required={false}
                  tabIndex={-1}
                  value="title"
                />
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                  >
                    <span>
                      Sort by:
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              className="MuiButtonBase-root MuiIconButton-root iconButton"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            id="textField"
          >
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              onKeyDown={[Function]}
            >
              <label
                className="MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
                data-shrink={false}
                htmlFor="textField"
                id="textField-label"
              >
                Title
              </label>
              <div
                className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorSecondary MuiFilledInput-colorSecondary MuiInputBase-fullWidth MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiFilledInput-input"
                  disabled={false}
                  id="textField"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            id="searchBarButtonsContainer"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained searchBarButton MuiButton-containedPrimary"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Filters
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained searchBarButton MuiButton-containedPrimary"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Search
              </span>
            </button>
          </div>
        </div>
        <nav
          aria-label="pagination navigation"
          className="MuiPagination-root pagination"
        >
          <ul
            className="MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiPaginationItem-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-current="true"
                aria-label="page 1"
                className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-selected MuiPaginationItem-sizeLarge"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                1
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiPaginationItem-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
        <div
          id="results"
        >
          <p>
            Your query gave no results
          </p>
        </div>
        <nav
          aria-label="pagination navigation"
          className="MuiPagination-root pagination"
        >
          <ul
            className="MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiPaginationItem-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-current="true"
                aria-label="page 1"
                className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-selected MuiPaginationItem-sizeLarge"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                1
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiPaginationItem-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>,
  <div
    className="MuiBox-root MuiBox-root-7"
    id="footer"
  >
    <div
      id="footerContent"
    >
      <span>
        Made with love 
        &lt;3
      </span>
    </div>
  </div>,
]
`;
